### ISSUE TITLE
Migrate to the GitLab GraphQL API once the minimum version is 12.0

### ISSUE THREAD

**Issue by LawnGnome** [sourcegraph/sourcegraph#12167]
#11757 greatly expanded the calls we make to the GitLab API, particularly around merge requests. Those calls could be more efficiently made using the [GitLab GraphQL API](https://docs.gitlab.com/ee/api/graphql/getting_started.html): currently, we need to make three calls per MR to get the MR, notes, and pipelines; with GraphQL we could potentially make one call for an entire campaign's worth of MRs (subject to GitLab's own rate limiting), _and_ get better label information at the same time. (The REST API does not provide a way to get label descriptions or colours from a merge request.)

The GraphQL API only became available without enabling a feature flag in GitLab 12.0, so that would be the minimum version we would have to support before this becomes feasible.

**Comment by chrispine** [comment]
What versions do we currently support? (More importantly, how could I answer that question myself?)

**Comment by LawnGnome** [comment]
> What versions do we currently support?

Our minimum is [version 10](https://docs.sourcegraph.com/admin/external_service/gitlab#supported-versions).

**Comment by chrispine** [comment]
Ah, okay, so there's nothing on _our_ end that requires v10, just that that's what we support.

Do we have metrics on how many of our customers are on which versions? Who decides when we bump that number?

**Comment by chrispine** [comment]
yes, we do: https://sourcegraph.looker.com/dashboards/325?Unique+Server+ID=

**Comment by LawnGnome** [comment]
I discovered in #44692 that we actually need GitLab 14.1 for all the GraphQL fields that are required (specifically, `Pipeline.ref`, otherwise we could "only" require 13.7). I also discovered that there isn't a reasonable way to design an API with the right semantics that can delegate to both REST and GraphQL implementations — it ends up being a very leaky abstraction, or you just don't get the benefits of moving to GraphQL.

Since the REST code is well tested and will be sticking around anyway, I'm punting this one back to the long future. I expect we could reconsider this in 2024.
